### PR TITLE
New nodes

### PR DESCRIPTION
**Pull request by Potatooff**
Edit tab -> Undo, Redo button inside
seperator in file tab
Option menu -> Add node
New blur,
Modification on resize node

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -106,10 +107,19 @@ def duplicate_nodes(sender, app_data):
...
+    show=False,
+    label="Options",
+    width=250,
+    height=300,
 ):
-    for module in modules[1:-1]:
-        dpg.add_button(label=module.name, tag=module.name + "_popup", callback=module.run)
+    with dpg.menu(label="Add Node", indent=5):
+        for module in modules[1:-1]:
+            dpg.add_button(label=module.name, tag=module.name + "_popup", callback=module.run, indent=3, width=200)
```
Comment (FirePlank) [inline main.py:123]: we dont need a separate menu for now since we only have 1 functionality

File: main.py [hunk 2]
```diff
@@ -171,6 +181,12 @@ def handle_popup(_sender, app_data):
             dpg.add_menu_item(tag="export", label="Export image")
             dpg.add_menu_item(tag="close", label="Close app", callback=lambda: dpg.stop_dearpygui())
 
+        with dpg.menu(tag="Edit", label="Edit"):
+            dpg.add_button(label="Undo", tag="undo_button", enabled=False, width=150)
+            dpg.add_button(label="Redo", tag="redo_button", enabled=False, width=150)
+            dpg.add_separator()
+
+        with dpg.menu(tag="Nodes", label="Nodes"):
         with dpg.menu(tag="nodes", label="Nodes"):
```
Comment (FirePlank) [inline main.py:190]: why are you doing dpg.menu twice?

File: main.py [hunk 3]
```diff
@@ -157,7 +167,7 @@ def handle_popup(_sender, app_data):
                     width=200,
                     indent=5,
                 )
-
+            dpg.add_separator()
```
Comment (FirePlank) [inline main.py:170]: spacing between lines would be nice

File: src/corenodes/transform/blur.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    name = "Blur"
+    tooltip = "Blur area / full image"
+
+    def run(self):
+        with dpg.node(
+            parent="MainNodeEditor",
+            tag="blur_" + str(self.counter),
+            label="Blur image",
+            pos=[600, 20],
+        ):
```
Comment (FirePlank) [inline src/corenodes/transform/blur.py:17]: remember to set user_data to self

File: src/corenodes/transform/blur.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+            with dpg.node_attribute(attribute_type=dpg.mvNode_Attr_Output):
+                dpg.add_slider_int(
+                    tag="blur_percentage_" + str(self.counter),
+                    label="% Blur in percentage",
+                    width=150,
+                    default_value=0,
+                    max_value=100,
+                    min_value=0,
+                )
```
Comment (FirePlank) [inline src/corenodes/transform/blur.py:29]: why do we need 2 inputs to change the %?

File: src/corenodes/transform/blur.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+    def run(self):
+        with dpg.node(
+            parent="MainNodeEditor",
+            tag="blur_" + str(self.counter),
+            label="Blur image",
+            pos=[600, 20],
+        ):
+            with dpg.node_attribute():
+                dpg.add_input_int(tag="Blur_input" + str(self.counter), label="% Blur", width=100)
```
Comment (FirePlank) [inline src/corenodes/transform/blur.py:19]: remember that python uses snake_case

**Review by FirePlank — CHANGES_REQUESTED** [review]

**Comment by FirePlank** [comment]
Also renaming this PR to something else than test would be ideal.

File: main.py (hunk 1 above)
Comment (Potatooff) [reply main.py:123]: i know, i just added it so we dont forget it!

File: main.py (hunk 2 above)
Comment (Potatooff) [reply main.py:190]: I didn't noticed that.

File: main.py (hunk 3 above)
Comment (Potatooff) [reply main.py:170]: Okay got it

File: src/corenodes/transform/blur.py (hunk 5 above)
Comment (Potatooff) [reply src/corenodes/transform/blur.py:29]: its for flexibilty if user prefer giving an input or sliding the slider

**Pull request merged**